python/plugins/processing/algs/qgis/SplitWithLines.py

```diff
@@ -0,0 +1,175 @@
+        self.addParameter(ParameterVector(self.INPUT_B,
+                                          self.tr('Split layer'), [dataobjects.TYPE_VECTOR_LINE]))
+
+        self.addOutput(OutputVector(self.OUTPUT, self.tr('Splitted')))
+
+    def processAlgorithm(self, progress):
+        layerA = dataobjects.getObjectFromUri(self.getParameterValue(self.INPUT_A))
+        splitLayer = dataobjects.getObjectFromUri(self.getParameterValue(self.INPUT_B))
+
+        sameLayer = self.getParameterValue(self.INPUT_A) == self.getParameterValue(self.INPUT_B)
+        fieldList = layerA.fields()
+
+        writer = self.getOutputFromName(self.OUTPUT).getVectorWriter(fieldList,
+                                                                     layerA.wkbType(), layerA.crs())
+
+        spatialIndex = vector.spatialindex(splitLayer)
+        splitGeoms = {}
+        request = QgsFeatureRequest()
+        request.setSubsetOfAttributes([])
+
+        for aSplitFeature in vector.features(splitLayer, request):
+            splitGeoms[aSplitFeature.id()] = aSplitFeature.geometry()
+            # honor the case that user has selection on split layer and has setting "use selection"
+
+        outFeat = QgsFeature()
+        features = vector.features(layerA)
+
+        if len(features) == 0:
+            total = 100
+        else:
+            total = 100.0 / float(len(features))
+
+        multiGeoms = 0 # how many multi geometries were encountered
+
+        for current, inFeatA in enumerate(features):
+            inGeom = inFeatA.geometry()
+
+            if inGeom.isMultipart():
+                multiGeoms += 1
+                # MultiGeometries are not allowed because the result of a splitted part cannot be clearly defined:
+                # 1) add both new parts as new features
+                # 2) store one part as a new feature and the other one as part of the multi geometry
+                # 2a) which part should be which, seems arbitrary
+            else:
+                attrsA = inFeatA.attributes()
+                outFeat.setAttributes(attrsA)
+                inGeoms = [inGeom]
+                lines = spatialIndex.intersects(inGeom.boundingBox())
+
+                if len(lines) > 0:  # has intersection of bounding boxes
+                    splittingLines = []
+
+                    for i in lines:
+                        try:
+                            splitGeom = splitGeoms[i]
+                        except:
+                            continue
+
+                        # check if trying to self-intersect
+                        if sameLayer:
+                            if inFeatA.id() == i:
+                                continue
+
+                        if inGeom.intersects(splitGeom):
+                            splittingLines.append(splitGeom)
+
+                    if len(splittingLines) > 0:
+                        for splitGeom in splittingLines:
+                            splitterPList = vector.extractPoints(splitGeom)
```

Comment:
you could move this after the intersects test in line 126 - if there's no intersection then you'd avoid the cost of extracting the unused points